### PR TITLE
avoid create-or-replace of deployment when scaling

### PR DESCRIPTION
**Pull request by markfisher**

### CODE REVIEW

File: function-controller/src/main/java/io/projectriff/controller/function/FunctionDeployer.java [hunk 1]
```diff
@@ -66,11 +66,12 @@ public FunctionDeployer(KubernetesClient kubernetesClient) {
...
 	 */
-	public void deploy(XFunction functionResource, int replicas) {
+	public void deploy(XFunction functionResource) {
 		String functionName = functionResource.getMetadata().getName();
-		logger.debug("Setting {} replicas for {}", replicas, functionName);
+		int replicas = 0; // TODO: allow configuration of minReplicas for a function?
+		logger.debug("Deploying {} with {} replicas", functionName, replicas);
 		// @formatter:off
 		this.kubernetesClient.extensions().deployments()
 				.inNamespace(functionResource.getMetadata().getNamespace())
```
Comment (ericbottard) [inline function-controller/src/main/java/io/projectriff/controller/function/FunctionDeployer.java]: Should the next line be changed from `createOrReplaceWithNew()` to whatever is appropriate to just "create"?

Comment (markfisher) [reply function-controller/src/main/java/io/projectriff/controller/function/FunctionDeployer.java]: good catch, just pushed that change

**Comment by ericbottard** [comment]
While I understand the rationale for this change, thinking about it some more I think we should try to fix the "race condition" rather than change from a "describe what state we want" (which we have currently) to a "describe the changes we we think we need to reach the wanted state".

One of the reasons I'm saying this, is the following:

If the controller crashes and restarts, it will read the current CRD state of topics and functions and fire fake "create" events. With your changes this will fire `createNew` API calls, but the deployments are already there!

**Comment by ericbottard** [comment]
In other words I think this is `kubectl apply` vs. `kubectl create`, and my understanding is that we want to strive for the former

**Comment by markfisher** [comment]
I understand the point for the restart/replay situation, but I don't want to revert to `createOrReplaceWithNew` for every scale operation. IIUC simply updating the replica count via `scale` is far lighter (and we need to be as light/responsive as possible for scaling). In fact, I think the `apply` vs `create` analogy is the opposite of what you're suggesting, since we are currently *creating* (replacing) every time, whereas an actual `kubectl apply` on a `Deployment` with a different replica count would simply "scale".

**Comment by markfisher** [comment]
for the replay of fake "added" events when a Deployment already exists (e.g. after deleting the `function-controller` pod and letting it restart automatically), this currently happens on the PR branch:

```[function-controller-597283684-49lf9] io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://kubernetes.default.svc/apis/extensions/v1beta1/namespaces/default/deployments. Message: deployments.extensions "square" already exists. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=extensions, kind=deployments, name=square, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=deployments.extensions "square" already exists, metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}), reason=AlreadyExists, status=Failure, additionalProperties={}).```

So... we definitely do need to address that, but I think ignoring it and logging that it already exists might be sufficient. WDYT @ericbottard?

**Comment by ericbottard** [comment]
I think it's fine, this was just an immediate example of what I think could go wrong.

I'd be interested in learning more about the actual implementation of `kubectl apply` vs `kubectl create` though. Is it your understanding that `apply` does a diff clientside?

**Comment by markfisher** [comment]
https://kubernetes.io/docs/tutorials/object-management-kubectl/declarative-object-management-configuration/#how-apply-calculates-differences-and-merges-changes

also in that doc:

> Directly update the replicas field in the live configuration by using kubectl scale. This does not use kubectl apply:

```
kubectl scale deployment/nginx-deployment --replicas=2
```

**Comment by markfisher** [comment]
I pushed a commit to handle (ignore) the already-exists case on restart/replay.

**Comment by ericbottard** [comment]
Merged as 3c4def8

**Comment by ericbottard** [comment]
Not the subject of this PR, but I played a bit with `kubectl delete` while this was running:

Shouldn't a `deployment` be recreated when a `function` for it exists? That is, in the same way that `k delete po/foo-xxx` ends up triggering the recreation of a new `pod` as long as there is a `deployment`, shouldn't `k delete deploy/square` end up being forcibly recreated as long as the `function` for it exists? @markfisher

**Comment by markfisher** [comment]
Interesting question. I suppose if we consider the `Deployment` to be an implementation detail of how we deploy `Functions`, then yes. If someone intervenes manually, we should be able to restore the correct state (as long as the corresponding `Function` does exist). If so, I think this should be handled by catching the exception for a `scale()` call when the underlying `Deployment` no longer exists. In other words, we should still rely on just `scale()` for the "normal" case.

**Comment by markfisher** [comment]
I like that better. We are already watching `Deployment` events, so if we see a delete but a `Function` does still exist for that `Deployment`, we can call our `FunctionDeployer.deploy()`. Scaling will kick in from there... so at least the manual deletion does reset the replica count to 0 thus "resetting" the Function overall, as deleting a Pod triggers a reset (new instance).

**Pull request closed without merging**